### ISSUE TITLE
IX_PSAM_AllowFiltering is slowing down queries

### ISSUE THREAD

**Issue by claymckinney** [nopSolutions/nopCommerce#3218]
nopCommerce version: 4.0

Steps to reproduce the problem: 
We have hundreds of thousands of products and tens of thousands of categories, so slight performance issues are amplified for us. By removing this index, we were able to shave 12 seconds off the time to run ProductLoadAllPaged.

**Comment by PavelV95** [comment]
> Clarification: Running ProductLoadAllPaged with some criteria took 12 seconds before, and 0 seconds after we removed the index. Running it with NO criteria took 27 seconds before, and 3 seconds after we removed the index.

Please, give me your SQL query from the profiler

**Comment by skoshelev** [comment]
We conducted a series of tests and could not confirm the problem

Database:
Product - 38587 rows
Product_SpecificationAttribute_Mapping  - 176989 rows

Code from ProductLoadAllPaged procedure:

```
declare @FilterableSpecificationAttributeOptionIds nvarchar(MAX)
CREATE TABLE #FilterableSpecs 
		(
			[SpecificationAttributeOptionId] int NOT NULL
		)
        DECLARE @sql_filterableSpecs nvarchar(max)
        SET @sql_filterableSpecs = '
	        INSERT INTO #FilterableSpecs ([SpecificationAttributeOptionId])
	        SELECT DISTINCT [psam].SpecificationAttributeOptionId
	        FROM [Product_SpecificationAttribute_Mapping] [psam] WITH (NOLOCK)
	            WHERE [psam].[AllowFiltering] = 1
	            AND [psam].[ProductId] IN (SELECT [Id] FROM [Product])'

        EXEC sp_executesql @sql_filterableSpecs

		--build comma separated list of filterable identifiers
		SELECT @FilterableSpecificationAttributeOptionIds = COALESCE(@FilterableSpecificationAttributeOptionIds + ',' , '') + CAST(SpecificationAttributeOptionId as nvarchar(4000))
		FROM #FilterableSpecs

		DROP TABLE #FilterableSpecs

print @FilterableSpecificationAttributeOptionIds
```
Result with index
![with_index](https://user-images.githubusercontent.com/16875599/52411747-daeb9a00-2aed-11e9-8a01-74b643200498.jpg)
Result without index
![without_index](https://user-images.githubusercontent.com/16875599/52411781-f22a8780-2aed-11e9-9fdb-816d5698cbf4.jpg)

As we see the index help improve performance, albeit not very significantly, but it doesn't slow down the stored procedure in a fact

Closed #3218
